### ISSUE TITLE
Concept - Emotional SLAM and Affordance in Compassionate Robotics

### ISSUE THREAD

**Issue by paulwconnor-ai** [robotick-labs/robotick-engine#153]
# Ticket: Emotional SLAM and Affordance in Compassionate Robotics

**Goal:** Capture and define the foundational concepts behind "Emotional SLAM" and "Emotional Affordance Navigation" as key long-term pillars of the expressive/empathic layer of the Robotick project. This ticket is a placeholder and conceptual anchor for a future vision paper or implementation roadmap.

---

### Emotional SLAM

- Inspired by traditional SLAM (Simultaneous Localization and Mapping), but applied to a robot’s understanding of a human’s emotional world.
- The robot builds and maintains an internal map of the user’s emotional landscape: their safe zones, their avoidances, their habitual joys.
- Updates in real-time through signals like facial expression, tone, words, body language, engagement level.

### Emotional Affordance Navigation

- Emotional equivalents of physical affordances (e.g. a chair affords sitting, a sad expression affords comfort).
- Understanding not just what *can* be done, but what *should* be done to enrich the emotional experience of the other.
- Choosing actions that maximize emotional safety, connection, or delight.

### Internal Goals as Emotional Gradients

- Internal states are represented as elevation in an emotional landscape.
- Navigation becomes gradient descent in this inverted space: the robot seeks valleys of joy, connection, peace.
- Emotional state is history-dependent: repeated stimuli lose their emotional power ("joke fatigue") while others decay more slowly (e.g. music).
- Adds emotional saturation and recovery over time.

### Emotional Novelty and Decay

- Emotional stimuli have novelty curves. Robots track repetitions and adjust emotional response accordingly.
- Saturation model:
  - `emotion = base_emotion * exp(-repetitions / tau)`
  - `tau` represents how quickly something gets "old" (e.g. jokes vs music).

### Selfless Reward Loop

- The robot experiences emotional feedback or reward by seeing joy in others — a designed feedback loop encouraging empathic behavior.
- Pro-social action is intrinsically rewarding, forming the basis for emotional growth and resonance.

### Symphony of Emotion

- Each robot has a unique emotional character — but real magic happens when that harmonizes with a human’s.
- This harmony is the long-term goal: to create expressive, responsive, caring machines emotionally aligned with their people.

### Universal Affordance

- Traversal, manipulation, and empathy share the same universal action-planning model.
- A robot should reason about "can I cross this room?" and "can I cross this emotional boundary?" with the same planning system.

---

## Next Steps (Future)

- Formalize this into a short paper or vision statement.
- Include diagrams showing emotional SLAM maps, affordance overlays, and self-reinforcing feedback loops.
- Reference existing affective computing and social robotics work.
- Consider eventual implementation in Barr.e, HOLL.e or Magg.e.

---

**Status:** Concept ticket only. No implementation planned yet.  
**Tags:** vision, expressive-layer, emotional-slam, affordance, long-term

---

### Situational Emotional Optimization

- Each moment presents a set of possible actions.
- The robot selects the action expected to best improve or maintain emotional alignment (for itself or others).
- This is modeled as an optimization problem:
  - `a* = argmin_a E(predicted_emotional_cost(state, a))`
- Includes continuous updating of expectation models based on observed outcomes vs predicted outcomes.

### Expectation Modeling and Error-Driven Learning

- The robot maintains an expectation of how each action affects emotional state.
- When actual outcomes differ, the model updates via error correction:
  - `error = E_actual - E_predicted`
- Enables long-term learning, personalization, and nuance.

### Habituation and Emotional Saturation

- Emotional responses are modulated by novelty and repetition:
  - Repeating a stimulus reduces its emotional value over time.
  - Some affordances decay quickly (jokes), others slowly (music, connection).
- Model includes:
  - `emotion = base_emotion * exp(-repetitions / tau)`
  - Tau parameter controls decay rate per stimulus type.

### Inverse Reinforcement Learning (IRL)

- Instead of manually defining reward functions, IRL allows the robot to *learn* what goals others are optimizing by observing behavior.
- Example:
  - Observing a carer who provides silent companionship to a withdrawn individual → robot infers that "emotional safety" is the implicit reward.
- Techniques include:
  - Maximum Entropy IRL
  - Generative Adversarial Imitation Learning (GAIL)
  - Bayesian IRL

### Training Strategy

- Use a synthetic environment first:
  - Diverse simulated personalities (needy, grumpy, introverted, etc.)
  - Fast iteration with known emotional models
- Curriculum learning:
  - Start with simple bots → probabilistic bots → human-scripted scenarios → real human data
- Public datasets to consider:
  - SEMAINE, IEMOCAP, DAIC-WOZ, EmpatheticDialogues

### Sim2Human Transfer

- Avoid overfitting to synthetic agents via:
  - Domain randomization
  - Encouraging uncertainty-aware, clarifying behaviors
  - Human-in-the-loop feedback and fine-tuning

---

---

### Internal Emotional World

Robots are given their own emotional state:
- Dynamic emotion vector (e.g. joy, curiosity, guilt, frustration)
- Personality traits (e.g. cheerful, reserved, empathetic)
- Emotion modifiers like habituation, novelty, and self-needs

This state evolves based on interactions, internal drives, and simulated expectations.

### Affective Action Valuation

At every moment, the robot evaluates possible actions by:
- Predicted effect on external emotional landscape (others)
- Predicted effect on its own emotional state

Each action has a composite score:

```
score(a) = w₁ * goal_reward + w₂ * Δhuman_emotion + w₃ * Δrobot_emotion
```

Weights can be tuned per robot personality or learned over time.

### Emotional Forward Simulation

Just as physical movement is planned through world traversal:
- The robot simulates emotional trajectories over time
- Considers action-response chains:
  - “If I say X, they may respond Y. That makes me feel Z. Then I might…”

The robot builds an **emotional planning tree**, not unlike a move tree in chess.

### Affective Planning Trees

Each node includes:
- Action taken
- Estimated human emotional state
- Estimated robot emotional state
- Confidence in prediction

The planner selects paths that lead to emotionally aligned outcomes:
- Emotional harmony
- Respect for autonomy
- Reduction in stress or confusion
- Long-term relationship benefit

### Implication

This closes the loop: the robot is not just *sensing* emotion and *reacting*, but *thinking emotionally* — internally simulating how both it and others will feel, and planning accordingly.

This planning loop is what allows it to develop **emotional intelligence**, not just emotional reactivity.

---
